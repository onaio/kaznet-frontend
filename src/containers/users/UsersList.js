import React, { Component } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { bindActionCreators } from "redux";
import Moment from "react-moment";
import qs from "qs";
import "../LoadListAnimation.css";
import * as userActions from "../../store/users/actions";
import * as globalActions from "../../store/global/actions";
import * as globalSelectors from "../../store/global/reducer";
import * as userSelectors from "../../store/users/reducer";
import * as constants from "../../constants.js";
import NoResults from "../../components/NoResults";
import ListView from "../../components/ListView";
import ElementMap from "../ElementMap";

export class UsersList extends Component {
  async componentDidMount() {
    this.props.showListTitle();
    this.props.changePageTitle("Users");
    this.props.changePageTitleButton("+ Create User");
    this.props.changePageTarget("/users/new");

    let { search } = qs.parse(this.props.location.search.slice(1));
    const { page } = qs.parse(this.props.location.search.slice(1));

    if (search === undefined) {
      search = "";
    }
    this.props.searchVal(search);
    let pageNumber = Number(page);

    if (isNaN(pageNumber)) {
      pageNumber = 1;
    }
    await this.props.fetchUsers(
      `${
        constants.API_ENDPOINT
      }/userprofiles/?search=${search}&page=${pageNumber}`
    );
    this.props.changePageNumber(pageNumber);
  }

  componentDidUpdate(prevProps) {
    let { search } = qs.parse(this.props.location.search.slice(1));
    if (search === undefined) {
      search = "";
    }
    const { page } = qs.parse(this.props.location.search.slice(1));
    if (Number(page) !== this.props.currentPage && !isNaN(page)) {
      const pageNumber = Number(page);
      this.props.fetchUsers(
        `${
          constants.API_ENDPOINT
        }/userprofiles/?search=${search}&page=${pageNumber}`
      );
      this.props.changePageNumber(pageNumber);
    }
  }

  render() {
    console.log(this.props);
    if (this.props.searchParam !== "" && this.props.userCount === null) {
      return this.renderLoading();
    }
    if (this.props.userCount === 0) {
      return <NoResults searchVal={this.props.searchParam} />;
    }
    if (this.props.rowsIdArray.length <= 0) return this.renderLoading();
    return (
      <div className="UsersList">
        <ListView
          renderHeaders={this.renderHeaders}
          rowsIdArray={this.props.rowsIdArray}
          rowsById={this.props.rowsById}
          renderRow={this.renderRow}
          endpoint={"users"}
          pageLinks={this.props.pageLinks}
          totalPages={this.props.totalPages}
          currentPage={this.props.currentPage}
          firstPage={this.props.firstPage}
          lastPage={this.props.lastPage}
          searchVal={this.props.searchParam}
        />
      </div>
    );
  }

  renderLoading() {
    return (
      <center>
        <div className="lds-ripple">
          <div />
          <div />
        </div>
      </center>
    );
  }

  renderHeaders() {
    const headerItems = [
      "Role",
      "Username",
      "Last Name",
      "First Name",
      "Submissions",
      "Approved %",
      "Last Active Date"
    ];
    return <ElementMap items={headerItems} HTMLTag="th" />;
  }

  renderRow(row) {
    const rowItems = [
      row.attributes.role_display,
      <Link to={`/users/${row.id}`} key="link_to">
        {row.attributes.ona_username}
      </Link>,
      row.attributes.last_name,
      row.attributes.first_name,
      row.attributes.submission_count,
      row.attributes.approval_rate * 100 + "%",
      <Moment key={row.id} format="DD-MM-YYYY">
        {row.attributes.last_login}
      </Moment>
    ];

    return <ElementMap items={rowItems} HTMLTag="td" />;
  }
}

// which props do we want to inject, given the global store state?
function mapStateToProps(state) {
  return {
    rowsById: userSelectors.getUsersById(state),
    rowsIdArray: userSelectors.getUsersIdArray(state),
    totalPages: userSelectors.getTotalPages(state),
    currentPage: userSelectors.getCurrentPage(state),
    pageLinks: userSelectors.getPageLinks(state),
    firstPage: userSelectors.getFirstPage(state),
    lastPage: userSelectors.getLastPage(state),
    searchParam: globalSelectors.getSearchValue(state),
    userCount: userSelectors.getTotalCount(state)
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      fetchUsers: userActions.fetchUsers,
      changePageNumber: userActions.changePageNumber,
      changePageTitle: globalActions.changePageTitle,
      changePageTitleButton: globalActions.changePageTitleButton,
      changePageTarget: globalActions.changePageTarget,
      showListTitle: globalActions.toggleDetailTitleOff,
      searchVal: globalActions.getSearchVal
    },
    dispatch
  );
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UsersList);
