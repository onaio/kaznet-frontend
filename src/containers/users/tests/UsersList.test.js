import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Router } from "react-router";
import createBrowserHistory from "history/createBrowserHistory";

import { UsersList } from "../UsersList";
import * as fixtures from "../../../store/users/tests/fixtures";
const history = createBrowserHistory();

describe("containers/users/UsersList", () => {
  it("renders without crashing", () => {
    shallow(
      <UsersList
        fetchUsers={function() {}}
        changePageNumber={function() {}}
        changePageTitle={function() {}}
        changePageTitleButton={function() {}}
        changePageNumber={function() {}}
        showListTitle={function() {}}
      />
    );
  });

  it("renders user list correctly", () => {
    const wrapper = mount(
      <Router history={history}>
        <UsersList
          fetchUsers={function() {}}
          changePageNumber={function() {}}
          changePageTitle={function() {}}
          changePageTitleButton={function() {}}
          changePageNumber={function() {}}
          showListTitle={function() {}}
          rowsById={fixtures.usersById}
          endpoint={"users"}
          rowsIdArray={fixtures.usersIdArray}
          pageLinks={fixtures.pageLinks}
          totalPages={fixtures.totalPages}
          currentPage={fixtures.currentPage}
        />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
