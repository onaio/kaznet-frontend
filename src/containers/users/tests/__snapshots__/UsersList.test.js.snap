// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/users/UsersList renders user list correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <UsersList
    changePageNumber={[Function]}
    changePageTarget={[Function]}
    changePageTitle={[Function]}
    changePageTitleButton={[Function]}
    currentPage={1}
    endpoint="userprofiles"
    fetchUsers={[Function]}
    firstPage={1}
    lastPage={2}
    location={
      Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      }
    }
    pageLinks={
      Object {
        "first": "http://localhost:8000/api/v1/userprofiles/?page=1",
        "last": "http://localhost:8000/api/v1/userprofiles/?page=1",
        "next": "http://localhost:8000/api/v1/userprofiles/?page=2",
        "prev": null,
      }
    }
    rowsById={
      Object {
        "1": Object {
          "attributes": Object {
            "amount_earned": "0 KES",
            "approval_rate": 0,
            "approved_submissions": 1,
            "avg_amount_earned": 0,
            "avg_approval_rate": 0,
            "avg_approved_submissions": 0,
            "avg_rejected_submissions": 0,
            "avg_submissions": 0,
            "created": "2018-06-19T16:03:10.184043+03:00",
            "email": "sol@admin.me",
            "expertise": "1",
            "first_name": "Davis",
            "gender": "1",
            "last_login": "2018-06-19T16:03:22+03:00",
            "last_name": "Raymond",
            "modified": "2018-06-20T08:45:12.976916+03:00",
            "national_id": null,
            "ona_pk": null,
            "ona_username": "davisraym",
            "payment_number": "",
            "phone_number": "",
            "rejected_submissions": 0,
            "role": "1",
            "role_display": "Admin",
            "submission_count": 5,
          },
          "id": "1",
          "type": "UserProfile",
        },
      }
    }
    rowsIdArray={
      Array [
        "1",
      ]
    }
    searchParam=""
    searchVal={[Function]}
    showListTitle={[Function]}
    taskStatus=""
    totalPages={2}
  >
    <div
      className="UsersList"
    >
      <ListView
        currentPage={1}
        downloadModalHandler={[Function]}
        endpoint="users"
        firstPage={1}
        lastPage={2}
        modalState={false}
        onFormSubmit={[Function]}
        pageLinks={
          Object {
            "first": "http://localhost:8000/api/v1/userprofiles/?page=1",
            "last": "http://localhost:8000/api/v1/userprofiles/?page=1",
            "next": "http://localhost:8000/api/v1/userprofiles/?page=2",
            "prev": null,
          }
        }
        renderHeaders={[Function]}
        renderRow={[Function]}
        rowsById={
          Object {
            "1": Object {
              "attributes": Object {
                "amount_earned": "0 KES",
                "approval_rate": 0,
                "approved_submissions": 1,
                "avg_amount_earned": 0,
                "avg_approval_rate": 0,
                "avg_approved_submissions": 0,
                "avg_rejected_submissions": 0,
                "avg_submissions": 0,
                "created": "2018-06-19T16:03:10.184043+03:00",
                "email": "sol@admin.me",
                "expertise": "1",
                "first_name": "Davis",
                "gender": "1",
                "last_login": "2018-06-19T16:03:22+03:00",
                "last_name": "Raymond",
                "modified": "2018-06-20T08:45:12.976916+03:00",
                "national_id": null,
                "ona_pk": null,
                "ona_username": "davisraym",
                "payment_number": "",
                "phone_number": "",
                "rejected_submissions": 0,
                "role": "1",
                "role_display": "Admin",
                "submission_count": 5,
              },
              "id": "1",
              "type": "UserProfile",
            },
          }
        }
        rowsIdArray={
          Array [
            "1",
          ]
        }
        searchVal=""
        setUserDetails={[Function]}
        totalPages={2}
        userName={null}
      >
        <div>
          <ExportModal
            downloadModalHandler={[Function]}
            modalState={false}
            name={null}
            onFormSubmit={[Function]}
          >
            <div>
              <Modal
                autoFocus={true}
                backdrop={true}
                backdropTransition={
                  Object {
                    "mountOnEnter": true,
                    "timeout": 150,
                  }
                }
                centered={false}
                fade={true}
                isOpen={false}
                keyboard={true}
                modalTransition={
                  Object {
                    "timeout": 300,
                  }
                }
                onClosed={[Function]}
                onOpened={[Function]}
                role="dialog"
                toggle={[Function]}
                zIndex={1050}
              />
            </div>
          </ExportModal>
          <Table
            bordered={true}
            className="kaznet-table"
            responsiveTag="div"
            tag="table"
          >
            <table
              className="kaznet-table table table-bordered"
            >
              <thead>
                <tr>
                  <ElementMap
                    HTMLTag="th"
                    items={
                      Array [
                        "Role",
                        "Username",
                        "Last Name",
                        "First Name",
                        "Submissions",
                        "Approved %",
                        "Last Active Date",
                      ]
                    }
                  >
                    <th
                      key="0"
                    >
                      Role
                    </th>
                    <th
                      key="1"
                    >
                      Username
                    </th>
                    <th
                      key="2"
                    >
                      Last Name
                    </th>
                    <th
                      key="3"
                    >
                      First Name
                    </th>
                    <th
                      key="4"
                    >
                      Submissions
                    </th>
                    <th
                      key="5"
                    >
                      Approved %
                    </th>
                    <th
                      key="6"
                    >
                      Last Active Date
                    </th>
                  </ElementMap>
                </tr>
              </thead>
              <tbody>
                <tr
                  key="1"
                >
                  <ElementMap
                    HTMLTag="td"
                    items={
                      Array [
                        "Admin",
                        <Link
                          replace={false}
                          to="/users/1"
                      >
                          davisraym
                      </Link>,
                        "Raymond",
                        "Davis",
                        <div>
                          5
                          <Button
                              className="mx-2 btn btn-sm btn-light white"
                              color="secondary"
                              onClick={[Function]}
                              tag="button"
                          >
                              <FontAwesomeIcon
                                  border={false}
                                  className="fa-xs icon-link withspace"
                                  fixedWidth={false}
                                  flip={null}
                                  icon="laptop"
                                  listItem={false}
                                  mask={null}
                                  name=""
                                  pull={null}
                                  pulse={false}
                                  rotation={null}
                                  size={null}
                                  spin={false}
                                  symbol={false}
                                  transform={null}
                              />
                          </Button>
                      </div>,
                        "0%",
                        <t
                          ago={false}
                          calendar={false}
                          decimal={false}
                          element="time"
                          filter={[Function]}
                          format="DD-MM-YYYY"
                          fromNow={false}
                          interval={60000}
                          onChange={[Function]}
                          toNow={false}
                          unit={null}
                          unix={false}
                          utc={false}
                      >
                          2018-06-19T16:03:22+03:00
                      </t>,
                      ]
                    }
                  >
                    <td
                      key="0"
                    >
                      Admin
                    </td>
                    <td
                      key="1"
                    >
                      <Link
                        key="link_to"
                        replace={false}
                        to="/users/1"
                      >
                        <a
                          href="/users/1"
                          onClick={[Function]}
                        >
                          davisraym
                        </a>
                      </Link>
                    </td>
                    <td
                      key="2"
                    >
                      Raymond
                    </td>
                    <td
                      key="3"
                    >
                      Davis
                    </td>
                    <td
                      key="4"
                    >
                      <div
                        key="1"
                      >
                        5
                        <Button
                          className="mx-2 btn btn-sm btn-light white"
                          color="secondary"
                          onClick={[Function]}
                          tag="button"
                        >
                          <button
                            className="mx-2 btn btn-sm btn-light white btn btn-secondary"
                            onClick={[Function]}
                            type="button"
                          >
                            <FontAwesomeIcon
                              border={false}
                              className="fa-xs icon-link withspace"
                              fixedWidth={false}
                              flip={null}
                              icon="laptop"
                              listItem={false}
                              mask={null}
                              name=""
                              pull={null}
                              pulse={false}
                              rotation={null}
                              size={null}
                              spin={false}
                              symbol={false}
                              transform={null}
                            >
                              <svg
                                aria-hidden="true"
                                className="svg-inline--fa fa-laptop fa-w-20 fa-xs icon-link withspace"
                                data-icon="laptop"
                                data-prefix="fas"
                                role="img"
                                style={Object {}}
                                viewBox="0 0 640 512"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M512 64v256H128V64h384m16-64H112C85.5 0 64 21.5 64 48v288c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V48c0-26.5-21.5-48-48-48zm100 416H389.5c-3 0-5.5 2.1-5.9 5.1C381.2 436.3 368 448 352 448h-64c-16 0-29.2-11.7-31.6-26.9-.5-2.9-3-5.1-5.9-5.1H12c-6.6 0-12 5.4-12 12v36c0 26.5 21.5 48 48 48h544c26.5 0 48-21.5 48-48v-36c0-6.6-5.4-12-12-12z"
                                  fill="currentColor"
                                  style={Object {}}
                                />
                              </svg>
                            </FontAwesomeIcon>
                          </button>
                        </Button>
                      </div>
                    </td>
                    <td
                      key="5"
                    >
                      0%
                    </td>
                    <td
                      key="6"
                    >
                      <t
                        ago={false}
                        calendar={false}
                        decimal={false}
                        element="time"
                        filter={[Function]}
                        format="DD-MM-YYYY"
                        fromNow={false}
                        interval={60000}
                        key="1"
                        onChange={[Function]}
                        toNow={false}
                        unit={null}
                        unix={false}
                        utc={false}
                      >
                        <time
                          dateTime={"2018-06-19T13:03:22.000Z"}
                        >
                          19-06-2018
                        </time>
                      </t>
                    </td>
                  </ElementMap>
                </tr>
              </tbody>
            </table>
          </Table>
          <Pagination
            aria-label="Page navigation example"
            listTag="ul"
            tag="nav"
          >
            <nav
              aria-label="Page navigation example"
              className=""
            >
              <ul
                className="pagination"
              >
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="First"
                      className="page-link"
                      replace={false}
                      to="/users/?search=&status=&page=1"
                    >
                      <a
                        aria-label="First"
                        className="page-link"
                        href="/users/?search=&status=&page=1"
                        onClick={[Function]}
                      >
                        First
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={true}
                  tag="li"
                >
                  <li
                    className="page-item disabled"
                  >
                    <Link
                      aria-label="Previous"
                      className="page-link"
                      replace={false}
                      to="#"
                    >
                      <a
                        aria-label="Previous"
                        className="page-link"
                        href="/"
                        onClick={[Function]}
                      >
                        «
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="Next"
                      className="page-link"
                      replace={false}
                      to="/users/?search=&status=&page=2"
                    >
                      <a
                        aria-label="Next"
                        className="page-link"
                        href="/users/?search=&status=&page=2"
                        onClick={[Function]}
                      >
                        »
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="Last"
                      className="page-link"
                      replace={false}
                      to="/users/?search=&status=&page=2"
                    >
                      <a
                        aria-label="Last"
                        className="page-link"
                        href="/users/?search=&status=&page=2"
                        onClick={[Function]}
                      >
                        Last
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
              </ul>
            </nav>
          </Pagination>
        </div>
      </ListView>
    </div>
  </UsersList>
</Router>
`;
